torchrl/modules/models/model_based_models.py

```diff
@@ -0,0 +1,316 @@
+            obs_embedding: a batch x time_steps x latent_size tensor containing the sequence of encoded observations
+
+        Returns:
+            prior_means: a batch x time_steps x state_size containing the mean of the state distributions
+            prior_stds: a batch x time_steps x state_size containing the standard deviation of the state distributions
+            prior_states: a batch x time_steps x state_size containing the sampled states
+            beliefs: a batch x time_steps x belief_size containing the sequence of beliefs
+            posterior_means: a batch x time_steps x state_size containing the mean of the posterior state distributions
+            posterior_stds: a batch x time_steps x state_size containing the standard deviation of the posterior state distributions
+            posterior_states: a batch x time_steps x state_size containing the sampled posterior states
+        """
+        prior_means = []
+        prior_stds = []
+        prior_states = []
+        beliefs = []
+        posterior_means = []
+        posterior_stds = []
+        posterior_states = []
+
+        for i in range(actions.shape[1]):
+            prior_mean, prior_std, prior_state, belief = self.rssm_prior(
+                posterior_state, belief, actions[:, i]
+            )
+            posterior_mean, posterior_std, posterior_state = self.rssm_posterior(
+                belief, obs_embedding[:, i]
+            )
+            prior_means.append(prior_mean)
+            prior_stds.append(prior_std)
+            prior_states.append(prior_state)
+            beliefs.append(belief)
+            posterior_means.append(posterior_mean)
+            posterior_stds.append(posterior_std)
+            posterior_states.append(posterior_state)
+        prior_means = torch.stack(prior_means, dim=1)
+        prior_stds = torch.stack(prior_stds, dim=1)
+        prior_states = torch.stack(prior_states, dim=1)
+        beliefs = torch.stack(beliefs, dim=1)
+        posterior_means = torch.stack(posterior_means, dim=1)
+        posterior_stds = torch.stack(posterior_stds, dim=1)
+        posterior_states = torch.stack(posterior_states, dim=1)
+        return (
+            prior_means,
+            prior_stds,
+            prior_states,
+            beliefs,
+            posterior_means,
+            posterior_stds,
+            posterior_states,
+        )
+
+
+class RSSMPrior(nn.Module):
+    """The prior network of the RSSM.
+
+    This network takes the previous state and belief and the action and outputs the next state and belief.
+    State is by construction stochastic and belief is deterministic. In the paper this is called the
+    deterministic state and stochastic state respectively. We prefer to call them belief and state for clarity.
+    Furthermore, the paper refers to both prior and posterior as state. We will disntinguish them by calling
+    the prior state and posterior state.
```

Comment:
`We distinguish states sampled according prior and posterior distribution for clarity.`